ui: bazel build for UI assets (db-console and cluster-ui)

koorosh opened the pull request:
- build `cluster-ui` JS package
- build `db-console` package
- build `crdb-protobufjs-client` package
- embed JS assets into `distccl` go's binary

Depends on PR: https://github.com/cockroachdb/cockroach/pull/69135

Release justification: non-production code changes

koorosh committed:
ui: bazel build for UI

- build `cluster-ui` JS package
- build `db-console` package
- build `crdb-protobufjs-client` package
- embed JS assets into `distccl` go's binary

Release note: None

Release justification: non-production code changes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67066)
<!-- Reviewable:end -->

koorosh commented on pkg/ui/BUILD.bazel:
```diff
@@ -13,3 +13,17 @@ go_library(
     ],
 )
+
+filegroup(
```
get all `node_module`'s packages except local linked packages

rickystewart replied:
`node_modules` is a `gitignore`d directory, so this seems wrong. What are you trying to do here?

koorosh replied:
Bazel treats `node_modules` as an external repository and used to include external dependencies:
```
workspace(
    name = "cockroach",
    managed_directories = {
       "@npm": ["pkg/ui/node_modules"]
    },
)
```
except external packages, it also included symlinked local packages
```
|- workspaces
    |- cluster-ui
    |- db-console
|- node_modules
    |- ... 
    |- cluster-ui (symlink) 
```
and to avoid circular dependencies for cluster-ui rules, we need all packages from node_modules except `cluster-ui`.

rickystewart commented:
bazel side looks fine so far, i'll leave the remainder of the review to someone who's knowledgeable about UI stuff.

koorosh commented:
@rickystewart , can you help me to resolve following issue with CI build? It fails due to error with writing files to filesystem:
> error Could not write file "/go/src/github.com/cockroachdb/cockroach/pkg/ui/yarn-error.log": "EROFS: read-only file system, open '/go/src/github.com/cockroachdb/cockroach/pkg/ui/yarn-error.log'"
  error An unexpected error occurred: "EROFS: read-only file system, mkdir '/go/src/github.com/cockroachdb/cockroach/pkg/ui/node_modules'".

I've managed to resolve the same problem locally by adding `rw` permissions to mounted docker volume (for `dev builder`) (https://github.com/cockroachdb/cockroach/pull/67066/commits/14d34ab9a4f783f83c5a89df3ed14f6f8defe5a6) but it didn't help on Teamcity.

This issue happens because bazel installs `node_modules` in host environment and then links it inside of sandboxed env.

rickystewart commented:
@koorosh Good catch, I'll send out a PR to fix that in a moment.

rickystewart commented:
https://github.com/cockroachdb/cockroach/pull/69643

rickystewart commented on WORKSPACE:
```diff
@@ -2,7 +2,9 @@
     name = "cockroach",
-    managed_directories = {"@npm": ["node_modules"]},
+    managed_directories = {
+       "@npm": ["pkg/ui/node_modules"]
```
Do the other `node_modules` directories (`pkg/ui/workspaces/cluster-ui/node_modules`, `pkg/ui/workspaces/db-console/node_modules`) need to be in this list as well?

koorosh replied:
Done, thanks!

rickystewart commented on pkg/ui/distccl/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+    name = "assets",
+    srcs = ["//pkg/ui/workspaces/db-console:db_console_webpack"],
+    outs = ["assets"],
+    cmd = "cp -r $(locations //pkg/ui/workspaces/db-console:db_console_webpack)/* $(OUTS)",
```
Nit: `-R` is more standard than `-r` here.

koorosh replied:
Done

rickystewart commented on pkg/ui/distccl/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+genrule(
+    name = "assets",
```
Bad style for a rule to have the same name as an output file, so you should give the rule a different name (e.g. `genassets`).

koorosh replied:
Done

rickystewart commented on pkg/ui/distccl/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+genrule(
+    name = "assets",
+    srcs = ["//pkg/ui/workspaces/db-console:db_console_webpack"],
+    outs = ["assets"],
```
So generally you [should](https://docs.bazel.build/versions/main/be/general.html#general-advice) avoid creating directories as `genrule` outputs:

> Avoid creating symlinks and directories. Bazel doesn't copy over the directory/symlink structure created by genrules and its dependency checking of directories is unsound.

And indeed when I run the build I get this warning trying to build this target:

    WARNING: /cockroach/pkg/ui/distccl/BUILD.bazel:3:8: output 'pkg/ui/distccl/assets' of //pkg/ui/distccl:genassets is a directory; dependency checking of directories is unsound

For *now*, if we have to do this, this is fine and we can check it in so we can unblock other work, but we would want to fix this before we can call this production-ready. A couple options:

1. You can declare all the sub-files under the `assets` directory explicitly. I see in that directory there are a lot of files with long unpredictable names (like `02290aae2a70e2d8dc8d391049661b25.woff2`), so maybe this isn't possible? 
2. You can compress the `assets` directory into a single file (like a `.tar.gz`), and then in the Go code you can decompress the `.tar.gz` and expose those artifacts as an `FS`. [Very small example.](https://stackoverflow.com/questions/51073133/bazel-genrule-that-outputs-a-directory)

(2) is probably what you'll have to do, although it'll take a little refactoring on the Go side. Like I said, we can take that as a TODO for later.

rickystewart replied:
Oh, there might be a low-friction third option that allows us to accomplish the same goal without a bunch of refactoring. Not really sure, I'd have to look closer. Happy to have a call about this if you want to.

rickystewart commented on pkg/ui/distoss/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+genrule(
+    name = "assets",
```
Same comments as above: should have distinct `name`, directory output is unsound, use `cp -R` instead of `cp -r`

koorosh replied:
Done

rickystewart commented on pkg/ui/workspaces/db-console/src/app.tsx:
```diff
@@ -14,7 +14,9 @@ import "nvd3/build/nv.d3.min.css";
-import "react-select/dist/react-select.css";
+// TODO (koorosh): commented out to make bazel build pass
+// TODO (koorosh): cluster-ui and db-console depend on differend versions of react-select
+// import "react-select/dist/react-select.css";
```
Does the UI build (as built with `make`) still work even with this commented out?

koorosh replied:
Oh, this should be fixed before merging this PR.

rickystewart commented on pkg/ui/workspaces/cluster-ui/webpack.config.js:
```diff
@@ -143,7 +144,9 @@ module.exports = {
-    "react-router-dom": "react-router-dom",
+    // TODO (koorosh): commented to avoid bazel build failure.
+    // should be fixed to consume react-router-dom as peer dependency
+    // "react-router-dom": "react-router-dom",
```
Same question -- can this safely be commented out without breaking the `make` build?

koorosh replied:
Done, I added condition in webpack configuration to exclude this change for bazel build only. Additional `is_bazel_build` env variable is passed to coordinate this flow.

rickystewart reviewed (changes requested):
I didn't look too closely at the JS changes, you'll want to get another reviewer to make sure those changes are sound.

rickystewart commented:
Oh, rebase atop `master` to get rid of those conflicts/fix the Docker mount issue.

koorosh committed:
bazel: fix npm dependencies resolution for bazel build

- include workspace specific dependencies to `db-console` build (for react-select library)

Release note: None

Release justification: non-production code changes

koorosh committed:
ui: restore babel.config.js required for tests

`ui-test` target fails without presence of babel.config.js
file, so it was restored.

Release note: None

Release justification: non-production code changes

koorosh committed:
ui: clean up bazel and webpack build configs

- rename main targets in bazel builds to match with
dir name for more convenient namings
- clean up webpack configs to separate configs which
are required for bazel builds only.

Release note: None

Release justification: non-production code changes

koorosh commented:
Updated current PR with following changes:
- webpack configuration for `db-console` and `cluster-ui` aren't changed for builds with `make`, and it's extended in a way to consume extra env variables to alter configuration required for Bazel
- fixed react router navigation for bazel build
- resolved conflicting dependency (`react-select`) between `cluster-ui` and `db-console`

koorosh committed:
ui: UI bazel build for CCL and OSS licence

Extend UI builds for CCL and OSS licences.

Release note: None

Release justification: non-production code changes

rickystewart reviewed (approved):
Works for me, we have to take care of that "genrule outputs a directory" thing though.

koorosh committed:
update build_bazel_rules_nodejs build rules to version 4.1.0

Use the latest version for build_bazel_rules_nodejs rules and update
version of @bazel/typescript package to match.

Release note: none

Release justification: non-production code changes

koorosh commented:
@rickystewart , the only change is added after your last review is version update for `build_bazel_rules_nodejs` (d366c83bfefeeb4ac6798f2407d749d7eb022792) and minor code formatting.

It is possible to build `cockroach` binary with following steps:
- start docker container with `dev builder`
- install `node_modules` as a self-managed external repository: `bazel run @nodejs//:yarn_node_repositories`
- build binary: `bazel build //pkg/cmd/cockroach` or `bazel build //pkg/cmd/cockroach-oss`

rickystewart commented on pkg/ui/workspaces/db-console/BUILD.bazel:
```diff
@@ -0,0 +1,90 @@
+load("@npm//webpack-cli:index.bzl", webpack = "webpack_cli")
```
Nit: put a blank line after this line (won't pass CI unless you do)

rickystewart commented on pkg/ui/distccl/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+# TODO (koorosh): genrule produces directory as output, should be changed later.
```
Is there a bug filed for this?

koorosh replied:
Thanks for reminder! 
https://github.com/cockroachdb/cockroach/issues/70348

rickystewart commented on WORKSPACE:
```diff
@@ -52,12 +59,21 @@ go_rules_dependencies()
     package_json = "//pkg/ui:package.json",
     yarn_lock = "//pkg/ui:yarn.lock",
+    strict_visibility = False,
+    # package_path = "pkg/ui/workspaces/a",
```
? Should be deleted?

koorosh replied:
Done

rickystewart reviewed (approved):
Works for me, let's get this landed so we can start playing with it.

koorosh committed:
build: code cleanup for bazel build files

Release note: none

Release justification: non-production code changes

koorosh commented:
bors r+
